homeassistant/components/crownstone/entry_manager.py

```diff
@@ -0,0 +1,226 @@
+                except UartException:
+                    delattr(self, ATTR_UART)
+                    # remove usb path to make usb setup available from options
+                    entry_data = self.config_entry.data.copy()
+                    entry_data[CONF_USB_PATH] = None
+                    self.hass.config_entries.async_update_entry(
+                        entry=self.config_entry, data=entry_data
+                    )
+                    # show notification to ensure the user knows the cloud is now used
+                    persistent_notification.async_create(
+                        self.hass,
+                        f"Setup of Crownstone USB dongle was unsuccessful on port {port}.\n \
+                        Crownstone Cloud will be used to switch Crownstones.\n \
+                        Please check if your port is correct and set up the USB again from integration options.",
+                        "Crownstone",
+                        "crownstone_usb_dongle_setup",
+                    )
+                # Create listeners for uart events
+                setup_uart_listeners(self)
+
+        # Save in what sphere the Crownstone USB is if it was setup correctly
+        # Crownstones could only be switched using the USB if they are in the same sphere (using BLE)
+        # Assuming that the other spheres are different buildings
+        if hasattr(self, ATTR_UART):
+            for sphere in self.cloud.cloud_data:
+                for crownstone in sphere.crownstones:
+                    if crownstone.type == CROWNSTONE_USB:
+                        self.usb_sphere_id = sphere.cloud_id
+
+        # create listener for when home assistant is stopped
+        self.hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, self.on_shutdown)
+
+        # register all entities
+        self.hass.config_entries.async_setup_platforms(self.config_entry, PLATFORMS)
+
+        return True
+
+    async def async_process_events(self, sse_client: CrownstoneSSEAsync) -> None:
+        """Asynchronous iteration of Crownstone SSE events."""
+        async with sse_client as client:
+            async for event in client:
+                if event is not None:
+                    # make SSE updates, like ability change, available to the user
+                    self.hass.bus.async_fire(f"{DOMAIN}_{event.type}", event.data)
+
+    async def async_unload(self) -> bool:
+        """Unload the current config entry."""
+        # authentication failed
+        if self.cloud.cloud_data is None:
+            return True
+
+        # close sse client and unsub from listeners
+        self.sse.close_client()
+        for sse_unsub in self.listeners[SSE_LISTENERS]:
+            sse_unsub()
+
+        # close uart connection and unsub from listeners
+        if hasattr(self, ATTR_UART):
+            self.uart.stop()
+            for subscription_id in self.listeners[UART_LISTENERS]:
+                UartEventBus.unsubscribe(subscription_id)
+
```

Comment:
Is this safe to run in the event loop.  Does it do I/O?

Reply:
The eventbus is a singleton https://github.com/crownstone/crownstone-lib-python-core/blob/master/crownstone_core/util/EventBus.py 

Initialized here https://github.com/crownstone/crownstone-lib-python-uart/blob/master/crownstone_uart/core/UartEventBus.py 

The unsubscribe function simply removes the callback from a dict. but from what i've seen the eventbus seems to run in one of the UartManager / UartBridge threads, but i don't maintain this one so it's hard to say.

Is it wise to wrap this function you think?

Reply:
Looks safe to run in the event loop assuming everything is threadsafe